// mysql2/typings/mysql/lib/protocol/sequences/promise/QueryableBase.ts
import {
  OkPacket,
  FieldPacket,
  RowDataPacket,
  ResultSetHeader,
} from 'mysql2/typings/mysql/lib/protocol/packets/index.js';
import { Query, QueryOptions, QueryableConstructor } from 'mysql2/typings/mysql/lib/protocol/sequences/Query.js';

export function QueryableBase<T extends QueryableConstructor>(
  Base: T = {} as T
) {
  return class extends Base {
    query<
      T extends
        | RowDataPacket[][]
        | RowDataPacket[]
        | OkPacket
        | OkPacket[]
        | ResultSetHeader
    >(sql: string): Promise<[T, FieldPacket[]]>;
    query<
      T extends
        | RowDataPacket[][]
        | RowDataPacket[]
        | OkPacket
        | OkPacket[]
        | ResultSetHeader
    >(
      sql: string,
      values: any | any[] | { [param: string]: any }
    ): Promise<[T, FieldPacket[]]>;
    query<
      T extends
        | RowDataPacket[][]
        | RowDataPacket[]
        | OkPacket
        | OkPacket[]
        | ResultSetHeader
    >(options: QueryOptions): Promise<[T, FieldPacket[]]>;
    query<
      T extends
        | RowDataPacket[][]
        | RowDataPacket[]
        | OkPacket
        | OkPacket[]
        | ResultSetHeader
    >(
      options: QueryOptions,
      values: any | any[] | { [param: string]: any }
    ): Promise<[T, FieldPacket[]]>;

    // Implementing all overload variations
    /* eslint-disable @typescript-eslint/no-unused-vars */
    query<
      T extends
        | RowDataPacket[][]
        | RowDataPacket[]
        | OkPacket
        | OkPacket[]
        | ResultSetHeader
    >(
      sql: string | QueryOptions,
      values?: any | any[] | { [param: string]: any }
    ): Promise<[T, FieldPacket[]]> {
      return new Promise(() => new Query());
    }
    /* eslint-enable @typescript-eslint/no-unused-vars */
  };
}
